File: runtime/orchestrator.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
from pathlib import Path
import sys
from typing import Any

sys.path.insert(0, str(Path(__file__).resolve().parent))

from orchestrator_core.protocol import ProtocolError  # noqa: E402
from orchestrator_core.store import WorkflowStore  # noqa: E402


def parser() -> argparse.ArgumentParser:
    result = argparse.ArgumentParser()
    result.add_argument("operation", choices=("next", "apply", "validate"))
    result.add_argument("--directory", required=True)
    result.add_argument("--request", required=True)
    return result


def body() -> dict[str, Any]:
    raw = sys.stdin.read()
    if not raw.strip():
        return {}
    value = json.loads(raw)
    if not isinstance(value, dict):
        raise ProtocolError("input", "must be a JSON object")
    return value


def main() -> int:
    args = parser().parse_args()
    try:
        payload = body()
        store = WorkflowStore(Path(args.directory), args.request)
        if args.operation == "next":
            state, action = store.reserve(expected_state_revision=payload.get("expected_state_revision"))
            result = {"ok": True, "state_revision": state["state_revision"], "status": state["status"], "action": action}
        elif args.operation == "apply":
            event = {
                "transition_id": payload.get("transition_id"),
                "type": payload.get("event_type"),
                "payload": payload.get("payload"),
            }
            state, applied = store.apply(event, expected_state_revision=payload.get("expected_state_revision"))
            result = {"ok": True, "state_revision": state["state_revision"], "status": state["status"], "result": applied}
        else:
            result = {"ok": True, "validation": store.validate()}
    except ProtocolError as error:
        result = {"ok": False, "error": {"type": "protocol", "field": error.field, "message": str(error), "value": error.value}}
    except Exception as error:
        result = {"ok": False, "error": {"type": "runtime", "message": str(error)}}
    sys.stdout.write(json.dumps(result, ensure_ascii=False, sort_keys=True))
    return 0 if result["ok"] else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/orchestrator.ts
import { tool, type ToolContext } from "@opencode-ai/plugin"
import { ProtocolError, WorkflowStore, parseJsonStrict, type EventInput } from "../runtime/orchestrator.js"

const requestSchema = tool.schema.string().min(1).max(80).regex(/^[a-z0-9][a-z0-9-]{0,79}$/)
const revisionSchema = tool.schema.number().int().nonnegative().optional()

function response(value: unknown): string {
  return JSON.stringify(value, null, 2)
}

function failure(error: unknown): string {
  if (error instanceof ProtocolError) {
    return response({ ok: false, error: { type: "protocol", field: error.field, message: error.message, value: error.value } })
  }
  return response({ ok: false, error: { type: "runtime", message: error instanceof Error ? error.message : String(error) } })
}

async function withStore<T>(request: string, context: ToolContext, operation: (store: WorkflowStore) => Promise<T>): Promise<string> {
  try {
    return response({ ok: true, ...(await operation(new WorkflowStore(context.directory, request)) as object) })
  } catch (error) {
    return failure(error)
  }
}

export const next = tool({
  description: "Reserve and return the single deterministic next technical-analysis action.",
  args: { request: requestSchema, expected_state_revision: revisionSchema },
  execute: (args: { request: string; expected_state_revision?: number }, context: ToolContext) => withStore(args.request, context, async (store) => {
    const { state, action } = await store.reserve(args.expected_state_revision)
    return { state_revision: state.state_revision, status: state.status, action }
  }),
})

export const apply = tool({
  description: "Apply one typed agent result or user decision to the pending technical-analysis transition.",
  args: {
    request: requestSchema,
    transition_id: tool.schema.string().min(1),
    event_type: tool.schema.string().min(1),
    payload_json: tool.schema.string().min(2),
    expected_state_revision: revisionSchema,
  },
  execute: (args: { request: string; transition_id: string; event_type: string; payload_json: string; expected_state_revision?: number }, context: ToolContext) => withStore(args.request, context, async (store) => {
    let payload: unknown
    try {
      payload = parseJsonStrict(args.payload_json)
    } catch (error) {
      throw new ProtocolError("payload_json", "invalid JSON", error instanceof Error ? error.message : String(error))
    }
    if (!payload || typeof payload !== "object" || Array.isArray(payload)) throw new ProtocolError("payload_json", "root must be an object")
    const event: EventInput = { transition_id: args.transition_id, type: args.event_type, payload: payload as Record<string, unknown> }
    const { state, result } = await store.apply(event, args.expected_state_revision)
    return { state_revision: state.state_revision, status: state.status, result }
  }),
})

export const validate = tool({
  description: "Validate technical-analysis state and artifacts without advancing the workflow.",
  args: { request: requestSchema },
  execute: (args: { request: string }, context: ToolContext) => withStore(args.request, context, async (store) => ({ validation: await store.validate() })),
})
